### PR TITLE
Multi-node LLM post-training with GRPO

### PR DESCRIPTION
**Pull request by pbelevich**

### CODE REVIEW

**Comment by KeitaW** [comment]
Thank you, @pbelevich. I will take a look at this PR over the next week!

File: 3.test_cases/pytorch/trl/grpo/train.sbatch [hunk 1]
```diff
@@ -0,0 +1,80 @@
+#!/bin/bash
+#SBATCH --job-name=grpo
+#SBATCH --nodes=9
+#SBATCH --ntasks-per-node 1
+
+## Set libfabric flags to use EFA
+export FI_PROVIDER=efa
+export FI_EFA_USE_DEVICE_RDMA=1 # use for p4d
```
Comment (KeitaW) [inline 3.test_cases/pytorch/trl/grpo/train.sbatch]:
```suggestion
```

File: 3.test_cases/pytorch/trl/grpo/train.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+            rewards.append(reward)
+        return rewards
+
+    def format_reward(completions, **kwargs):
+        """Reward function that checks if the completion has a specific format."""
+        pattern = r"^<think>.*?</think>\s*<answer>.*?</answer>$"
+        completion_contents = [completion[0]["content"] for completion in completions]
+        matches = [re.match(pattern, content) for content in completion_contents]
+        return [1.0 if match else 0.0 for match in matches]
+
```
Comment (KeitaW) [inline 3.test_cases/pytorch/trl/grpo/train.py:72]: Both of those two reward functions essentially checks we the completion contains `<think>.*?</think>\s*<answer>.*?</answer>$`, curious why we need to have both?

**Review by bluecrayon52 — COMMENTED** [review]
When testing with 5 p4d.24xlarge on HyperPod Slurm, using /fsx/ubuntu home directory, I had to run `mkdir ~/.cache/huggingface` then add `export HF_HOME=~/.cache/huggingface` to the `train.sbatch` file, otherwise it throws a formatting error for the `--container-mounts` flag, where the `HF_HOME` environment variable is referenced like so: `--container-mounts=.:/grpo,$HF_HOME:$HF_HOME`

File: 3.test_cases/pytorch/trl/grpo/train.py (hunk 2 above)
Comment (pbelevich) [reply 3.test_cases/pytorch/trl/grpo/train.py:72]: `simple_format_reward` gives partial rewards for finding the right tags and helps the model to reach perfect format which sometimes model cannot reach right away. So this helps the model to learn the right format faster instead of giving just 0 rewards for partially wrong format

File: 3.test_cases/pytorch/trl/grpo/README.md [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+```
+
+Install huggingface-cli:
+```bash
+pip install -U "huggingface_hub[cli]"
+```
+
+Download the model:
+```bash
+huggingface-cli download Qwen/Qwen2.5-72B
```
Comment (KeitaW) [inline 3.test_cases/pytorch/trl/grpo/README.md]:
```suggestion
hf download Qwen/Qwen2.5-72B
```
> ⚠️  Warning: 'huggingface-cli download' is deprecated. Use 'hf download' instead

File: 3.test_cases/pytorch/trl/grpo/README.md [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+sbatch train.sbatch Qwen/Qwen2.5-72B
+```
+
+The training script launches 8 nodes for training and 1 node for generation using vLLM, a high-throughput, low-latency inference engine for LLMs. The distributed training uses ZeRO stage 3 to accelerate the training process.
+
+The logs can be inspected using tail command:
+
+GRPO Training logs:
+```bash
+tail -f -n +0 grpo_XXX.out
```
Comment (KeitaW) [inline 3.test_cases/pytorch/trl/grpo/README.md]:
```suggestion
tail -f -n +0 grpo-XXX.out 
```

File: 3.test_cases/pytorch/trl/grpo/README.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+```
+sample output:
+```
+  1%|          | 17/2264 [01:22<2:55:16,  4.68s/it]
+0: {'loss': 0.0785, 'grad_norm': 0.8229517735973697, 'learning_rate': 9.916077738515903e-06, 'num_tokens': 1498339.0, 'completions/mean_length': 134.934765625, 'completions/min_length': 35.0, 'completions/max_length': 256.0, 'completions/clipped_ratio': 0.08203125, 'completions/mean_terminated_length': 124.83461303710938, 'completions/min_terminated_length': 35.0, 'completions/max_terminated_length': 253.8, 'rewards/format_reward/mean': 0.90703125, 'rewards/format_reward/std': 0.27258416190743445, 'rewards/accuracy_reward/mean': 0.224609375, 'rewards/accuracy_reward/std': 0.4104481041431427, 'reward': 1.131640625, 'reward_std': 0.34059175848960876, 'kl': 0.2958984375, 'clip_ratio/low_mean': 0.0, 'clip_ratio/low_min': 0.0, 'clip_ratio/high_mean': 0.0, 'clip_ratio/high_max': 0.0, 'clip_ratio/region_mean': 0.0, 'epoch': 0.01}
+```
+
+vLLM logs:
+```bash
+tail -f -n +0 vllm_XXX.out
```
Comment (KeitaW) [inline 3.test_cases/pytorch/trl/grpo/README.md]:
```suggestion
tail -f -n +0 vllm-XXX.out
```
